### PR TITLE
staticcheck: flag dubious bit shifting of fixed size integers

### PR DESCRIPTION
**Pull request by ksurent**
Fixes https://github.com/dominikh/go-tools/issues/654. Refers https://github.com/dominikh/go-tools/issues/149.

### CODE REVIEW

File: staticcheck/doc.go [hunk 1]
```diff
@@ -839,4 +839,27 @@ marshaling behavior, e.g. via MarshalJSON methods. It will also not
...
+
+	"SA9006": {
+		Title: `Dubious bit shifting of a fixed size integer value`,
+		Text: `Bit shifting an value past its size will always clear the value.
+
+For instance:
+
+	v := int8(42)
+	v >>= 8
+
```
Comment (dominikh) [inline staticcheck/doc.go:851]: Use 4 spaces instead of tabs for indentation, to be consistent with other examples.

File: staticcheck/doc.go [hunk 2]
```diff
@@ -839,4 +839,27 @@ marshaling behavior, e.g. via MarshalJSON methods. It will also not
...
+This check flags bit shifiting operations on fixed size integer values only.
+That is, int, uint and uintptr are never flagged to avoid potential false
+positives in somewhat exotic but valid bit twiddling tricks:
+
+	// Clear any value above 32 bits if integers are more than 32 bits.
+	func f(i int) int {
+		v := i >> 32
+		v = v << 32
+		return i-v
+	}`,
```
Comment (dominikh) [inline staticcheck/doc.go:863]: Four spaces

File: staticcheck/lint.go [hunk 3]
```diff
@@ -3817,3 +3819,58 @@ func CheckAddressIsNil(pass *analysis.Pass) (interface{}, error) {
...
 	return nil, nil
 }
+
+var (
+	checkFixedLengthTypeShiftQ = pattern.MustParse(`
+		(Or
+			(AssignStmt _ (Or ">>=" "<<=") _)
+			(BinaryExpr _ (Or ">>" "<<") _))
+	`)
+	wordSize = gcsizes.ForArch(runtime.GOARCH).WordSize
```
Comment (dominikh) [inline staticcheck/lint.go:3829]: Sizeof returns bytes, not words. WordSize returns the size of a word in bytes (4 bytes or 8 bytes), not bits in a byte. That makes `pass.TypesSizes.Sizeof(typ) * wordSize` incorrect.

File: staticcheck/lint.go [hunk 4]
```diff
@@ -3817,3 +3819,58 @@ func CheckAddressIsNil(pass *analysis.Pass) (interface{}, error) {
...
+			return 0, 0, false
+		}
+		switch typ.Kind() {
+		case types.Int8, types.Int16, types.Int32, types.Int64,
+			types.Uint8, types.Uint16, types.Uint32, types.Uint64:
+			// We're only interested in fixed–size types.
+		default:
+			return 0, 0, false
+		}
+		typeSize := pass.TypesSizes.Sizeof(typ) * wordSize
```
Comment (dominikh) [inline staticcheck/lint.go:3845]: Instead of figuring out the size of a byte, just assume that it is 8 bits.

File: staticcheck/lint.go [hunk 5]
```diff
@@ -3817,3 +3819,58 @@ func CheckAddressIsNil(pass *analysis.Pass) (interface{}, error) {
...
+	fn := func(node ast.Node) {
+		if _, ok := Match(pass, checkFixedLengthTypeShiftQ, node); !ok {
+			return
+		}
+
+		switch e := node.(type) {
+		case *ast.AssignStmt:
+			for i := range e.Lhs {
+				if size, shift, yes := isDubiousShift(e.Lhs[i], e.Rhs[i]); yes {
+					report.Report(pass, e, fmt.Sprintf("shifting %d bits value by %d will always clear it", size, shift))
```
Comment (dominikh) [inline staticcheck/lint.go:3864]: Let's try `shifting %d-bit value by %d bits will always clear it` instead

File: staticcheck/lint.go [hunk 6]
```diff
@@ -3817,3 +3819,58 @@ func CheckAddressIsNil(pass *analysis.Pass) (interface{}, error) {
...
+				}
+			}
+		case *ast.BinaryExpr:
+			if size, shift, yes := isDubiousShift(e.X, e.Y); yes {
+				report.Report(pass, e, fmt.Sprintf("shifting %d bits value by %d will always clear it", size, shift))
```
Comment (dominikh) [inline staticcheck/lint.go:3869]: Same

File: staticcheck/lint.go [hunk 7]
```diff
@@ -3817,3 +3819,58 @@ func CheckAddressIsNil(pass *analysis.Pass) (interface{}, error) {
...
+	}
+
+	fn := func(node ast.Node) {
+		if _, ok := Match(pass, checkFixedLengthTypeShiftQ, node); !ok {
+			return
+		}
+
+		switch e := node.(type) {
+		case *ast.AssignStmt:
+			for i := range e.Lhs {
```
Comment (dominikh) [inline staticcheck/lint.go:3862]: For assignment operators like `<<=`, there can only be one lhs and one rhs. (We only see ASTs that have already type-checked.)

File: staticcheck/doc.go [hunk 8]
```diff
@@ -839,4 +839,27 @@ marshaling behavior, e.g. via MarshalJSON methods. It will also not
 flag empty structs.`,
 		Since: "2019.2",
 	},
+
+	"SA9006": {
+		Title: `Dubious bit shifting of a fixed size integer value`,
+		Text: `Bit shifting an value past its size will always clear the value.
```
[the remark this answers is not in the document]
Comment (ksurent) [reply staticcheck/doc.go]: Fixed.

File: staticcheck/doc.go (hunk 2 above)
Comment (ksurent) [reply staticcheck/doc.go]: Fixed.

File: staticcheck/lint.go (hunk 5 above)
Comment (ksurent) [reply staticcheck/lint.go]: Fixed.

File: staticcheck/lint.go (hunk 6 above)
Comment (ksurent) [reply staticcheck/lint.go]: Fixed.

File: staticcheck/doc.go (hunk 1 above)
Comment (ksurent) [reply staticcheck/doc.go]: Fixed.

File: staticcheck/lint.go (hunk 7 above)
Comment (ksurent) [reply staticcheck/lint.go]: Fixed.

File: staticcheck/lint.go (hunk 3 above)
Comment (ksurent) [reply staticcheck/lint.go]: Fixed.

File: staticcheck/lint.go (hunk 4 above)
Comment (ksurent) [reply staticcheck/lint.go]: Fixed.

**Comment by ksurent** [comment]
Feedback addressed.
Not sure about the CI failure, tests pass on my laptop. Should I rebase onto the latest master, perhaps?

**Comment by dominikh** [comment]
You can rebase if you want to, but it will require some code changes, due to 9a1fe455724d08fc49829ac1e655fb2a0299d206. I'd be happy to rebase it for you.

**Comment by ksurent** [comment]
Please do!

**Comment by dominikh** [comment]
Merged as e306c661babf7e7dc9f0d784f83d820d7999c20e. Thank you!

**Pull request closed without merging**